core/services/workflows/syncer/workflow_registry.go

```diff
@@ -2,39 +2,552 @@ package syncer
+		ctx, cancel := w.stopCh.NewCtx()
+
+		w.wg.Add(1)
+		go func() {
+			defer w.wg.Done()
+			defer cancel()
+
+			w.syncEventsLoop(ctx)
+		}()
+
+		w.wg.Add(1)
+		go func() {
+			defer w.wg.Done()
+			defer cancel()
+
+			w.handlerLoop(ctx)
+		}()
+
+		return nil
+	})
 }
 
-func (w *WorkflowRegistry) Ready() error {
+func (w *workflowRegistry) Close() error {
+	return w.StopOnce(w.Name(), func() error {
+		close(w.stopCh)
+		w.wg.Wait()
+		return nil
+	})
+}
+
+func (w *workflowRegistry) Ready() error {
 	return nil
 }
 
-func (w *WorkflowRegistry) HealthReport() map[string]error {
+func (w *workflowRegistry) HealthReport() map[string]error {
 	return nil
 }
 
-func (w *WorkflowRegistry) Name() string {
-	return "WorkflowRegistrySyncer"
+func (w *workflowRegistry) Name() string {
+	return name
+}
+
+func (w *workflowRegistry) SecretsFor(ctx context.Context, workflowOwner, workflowName string) (map[string]string, error) {
+	return nil, errors.New("not implemented")
+}
+
+// handlerLoop handles the events that are emitted by the contract.
+func (w *workflowRegistry) handlerLoop(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case resp, open := <-w.eventsCh:
+			if !open {
+				return
+			}
+
+			if resp.Err != nil || resp.Event == nil {
+				w.lggr.Errorf("failed to handle event: %+v", resp.Err)
+				continue
+			}
+
+			event := resp.Event
+			w.lggr.Debugf("handling event: %+v", event)
+			if err := w.handler.Handle(ctx, *event); err != nil {
+				w.lggr.Errorf("failed to handle event: %+v", event)
+				continue
+			}
+		}
+	}
+}
+
+// syncEventsLoop polls the contract for events and passes them to a channel for handling.
+func (w *workflowRegistry) syncEventsLoop(ctx context.Context) {
+	var (
+		// sendLog is a helper that sends a WorkflowRegistryEventResponse to the eventsCh in a
+		// blocking way that will send the response or be canceled.
+		sendLog = func(resp WorkflowRegistryEventResponse) {
+			select {
+			case w.eventsCh <- resp:
+			case <-ctx.Done():
+			}
+		}
+
+		ticker = w.getTicker(ctx)
+
+		signal = make(chan struct{})
+	)
+
+	// critical failure if there is no reader, the loop will exit and the parent context will be
+	// canceled.
+	reader, err := w.getContractReader(ctx)
+	if err != nil {
+		w.lggr.Criticalf("contract reader unavailable : %s", err)
+		return
+	}
+
+	// fan out and query for each event type
+	for i := 0; i < len(w.eventTypes); i++ {
+		w.wg.Add(1)
+		go func() {
+			defer w.wg.Done()
+
+			queryEvent(
+				ctx,
+				signal,
+				w.lggr,
+				reader,
+				w.cfg,
+				w.eventTypes[i],
+				w.batchCh,
+			)
+		}()
+	}
+
+	// Periodically send a signal to all the queryEvent goroutines to query the contract
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker:
+			// for each event type, send a signal for it to execute a query and produce a new
+			// batch of event logs
+			for i := 0; i < len(w.eventTypes); i++ {
+				select {
+				case signal <- struct{}{}:
```

Comment:
this I think is a bug - there is no guarantee that the signal would not be consumed by just a few queryEvent loops and never seen by any of the other loops (in the extreme case), meaning  you would see a few event types multiple times, but not events of other types